sources/TileDB.CSharp/Query.cs

```diff
@@ -246,83 +220,430 @@ public Config Config()
+        /// </summary>
+        /// <typeparam name="T">The buffer's type.</typeparam>
+        /// <param name="name">The name of the attribute or the dimension.</param>
+        /// <param name="data">A <see cref="ReadOnlyMemory{T}"/> from where the data will be written.</param>
+        /// <exception cref="ArgumentException"><paramref name="data"/> is empty or <typeparamref name="T"/>
+        /// does not match the excepted data type.</exception>
+        /// <exception cref="NotSupportedException">The query's type is <see cref="QueryType.Read"/></exception>
+        public void SetDataWriteBuffer<T>(string name, ReadOnlyMemory<T> data) where T : struct
+        {
+            if (QueryType() == CSharp.QueryType.Read)
             {
-                throw new ArgumentException("Query.SetDataBuffer, buffer is null or empty!");
+                ThrowHelpers.ThrowOperationNotAllowedOnReadQueries();
             }
 
-            if (data is bool[] boolData && QueryType() == CSharp.QueryType.Write)
+            SetDataBuffer(name, MemoryMarshal.AsMemory(data));
+        }
+
+        /// <summary>
+        /// Sets the data buffer for an attribute or dimension
+        /// to a pinned memory buffer pointed by a <see cref="MemoryHandle"/>.
+        /// </summary>
+        /// <param name="name">The name of the attribute or the dimension.</param>
+        /// <param name="memoryHandle">A <see cref="MemoryHandle"/> pointing to the buffer.</param>
+        /// <param name="byteSize">The buffer's size <em>in bytes</em>.</param>
+        /// <exception cref="ArgumentException"><paramref name="byteSize"/> is zero.</exception>
+        /// <remarks>
+        /// <para>After calling this method, user code must not use <paramref name="memoryHandle"/>;
+        /// its ownership is transferred to this <see cref="Query"/> object.</para>
+        /// <para><paramref name="memoryHandle"/> will be disposed when one of the following happens:
+        /// <list type="bullet">
+        /// <item>The <see cref="Dispose"/> method is called.</item>
+        /// <item>The buffer for <paramref name="name"/> is reassigned through subsequent calls to this method.</item>
+        /// <item>This method call throws an exception.</item>
+        /// </list></para>
+        /// </remarks>
+        public void UnsafeSetDataBuffer(string name, MemoryHandle memoryHandle, ulong byteSize)
+        {
+            BufferHandle? handle = null;
+            bool successful = false;
+            try
+            {
+                handle = new BufferHandle(ref memoryHandle, byteSize);
+
+                SetDataBufferCore(name, handle.DataPointer, handle.SizePointer);
+
+                AddOrReplaceBufferHandle(_dataBufferHandles, name, handle);
+                successful = true;
+            }
+            finally
             {
-                SetDataBuffer<byte>(name, System.Array.ConvertAll(boolData as bool[], d => d ? (byte)1 : (byte)0));
```

Comment:
We might need to discuss this. The Core does not set any expectations about the definition of bool, particularly around the values that are not one or zero.

I removed it because there are correctness concerns besides performance ones. Imagine we set a data buffer to a `bool[]`, we submit the query and change our `bool[]` to contain the new data. _The Core will not see the new data._

And also since in C# we get bools that are not one or zero only by using `unsafe` code I don't think this is very impactful.